src/hotspot/share/runtime/arguments.cpp

```diff
@@ -3487,45 +3489,45 @@ void Arguments::extract_shared_archive_paths(const char* archive_path,
   len = end_ptr - begin_ptr;
   cur_path = NEW_C_HEAP_ARRAY(char, len + 1, mtInternal);
   strncpy(cur_path, begin_ptr, len + 1);
-  //cur_path[len] = '\0';
   FileMapInfo::check_archive((const char*)cur_path, false /*is_static*/);
   *top_archive_path = cur_path;
 }
 
-bool Arguments::init_shared_archive_paths() {
+void Arguments::init_shared_archive_paths() {
   if (ArchiveClassesAtExit != NULL) {
+    assert(!RecordDynamicDumpInfo, "already checked");
     if (DumpSharedSpaces) {
       vm_exit_during_initialization("-XX:ArchiveClassesAtExit cannot be used with -Xshare:dump");
     }
-    if (FLAG_SET_CMDLINE(DynamicDumpSharedSpaces, true) != JVMFlag::SUCCESS) {
-      return false;
-    }
     check_unsupported_dumping_properties();
-    SharedDynamicArchivePath = os::strdup_check_oom(ArchiveClassesAtExit, mtArguments);
```

Comment:
I think it is clearer to leave this assignment alone.
It works in your current patch because in line 3561, the `ArchiveClassesAtExit` is used instead of `SharedDynamicArchivePath`.

Reply:
SharedDynamicArchivePath had two meanings before this PR:

1. the dynamic archive that will be mapped during VM start-up. 
2. the dynamic archive that will be dumped.

This made the code complicated (E.g., the old version of DynamicArchive::dump() needed to call into Arguments::init_shared_archive_paths() to set it SharedDynamicArchivePath).

This PR removed the second meaning. That's why I removed line 3504. Now if ArchiveClassesAtExit is specified, SharedDynamicArchivePath will be NULL.

Reply:
Thanks for your explanation. It makes sense.